python-package/xgboost/core.py

```diff
@@ -427,21 +430,34 @@ def inner_f(*args, **kwargs):
     return inner_f
 
 
-class DMatrix:                  # pylint: disable=too-many-instance-attributes
+class DMatrix:  # pylint: disable=too-many-instance-attributes
     """Data Matrix used in XGBoost.
 
     DMatrix is an internal data structure that is used by XGBoost,
     which is optimized for both memory efficiency and training speed.
     You can construct DMatrix from multiple different sources of data.
     """
 
-    def __init__(self, data, label=None, weight=None, base_margin=None,
-                 missing=None,
-                 silent=False,
-                 feature_names=None,
-                 feature_types=None,
-                 nthread=None,
-                 enable_categorical=False):
+    @_deprecate_positional_args
+    def __init__(
+        self,
+        data,
+        label=None,
+        weight=None,
+        base_margin=None,
+        missing: Optional[float] = None,
+        *,
+        group=None,
+        qid=None,
+        label_lower_bound=None,
+        label_upper_bound=None,
+        silent: bool = False,
```

Comment:
Per my last remark about compatibility, please move this `slient` parameter to line 450 (in between `*` and `group`).

Reply:
The compatibility for `DMatrix` is kept.  But the `*` is still placed next to `label` to provide a warning for next release.  WDYT?

Reply:
Great! Thanks for keeping the order for the existing parameters.